reload: Check "/server_settings" before initiating a reload of the web-app.

laurynmm opened the pull request:
In order to avoid races with the user's device connecting to a network after unsuspending, we ping the `/server_settings` endpoint for a success response before initiating the reload of the app.

Fixes #33118.

---

**Notes**:
- I haven't able to reproduce the reported error/screen from the issue / CZO conversation in the dev environment, but I was able to trigger initiating a reload of the web-app in the dev environment.
- Currently, there isn't any `"#connection-error"` show/hide for `ui_report` included in these updates. I wasn't sure if going with something like we have in [`get_events`](https://github.com/zulip/zulip/blob/264e49e7e8ede23be94089b5dd2237f577ad8518/web/src/server_events.js#L148-L156) polling where there is a specific class added to the error message or with something more like [`message_fetch`](https://github.com/zulip/zulip/blob/264e49e7e8ede23be94089b5dd2237f577ad8518/web/src/message_fetch.ts#L384-L435) where there is no class added to the error message container.
- See in-line comments below ...

---

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [ ] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [ ] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [ ] Each commit is a coherent idea.
- [ ] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [ ] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.
</details>

laurynmm commented on web/src/reload.ts:
```diff
@@ -200,76 +203,137 @@ export function initiate({
-
     if (reload_state.is_pending() || reload_state.is_in_progress()) {
         return;
     }
```
It seemed better to check these before the check for a success response at `/server_settings`.

timabbott replied:
Hmm, but doing so would mean that the `immediate` flag doesn't take precedence over the in-progress state. I'm not sure whether that's important.

laurynmm replied:
Well, the `immediate` flag would be behind the call to `/server_settings`, so I thought it would be better to check whether a previous call to `reload.initiate` had toggled one of those two `reload_state` settings at the top.

timabbott replied:
I think this logic is wrong. If you look at the `server_events.js` caller to `reload.initiate`, it is very much expecting that this call will do something immediate if `reload_state.is_pending`:
```
                if (error_type !== "timeout" && reload_state.is_pending()) {                            
                    // The error might be due to the server changing                                    
                    reload.initiate({                                                                   
                        immediate: true,                                                                
                        save_compose: true,                                                             
                        send_after_reload: true,                                                        
                    });                                                                                 
                    return;                                                                             
                }                                                                                       
```

timabbott replied:
OK, the fix is to split up that if statement. The part inside `success` is what adds the pending state, so that's where we want the guard.

``` diff
diff --git a/web/src/reload.ts b/web/src/reload.ts
index 3c61389220..584aa883ea 100644
--- a/web/src/reload.ts
+++ b/web/src/reload.ts
@@ -203,7 +203,14 @@ export function initiate({
     send_after_reload = false,
     message_html = "Reloading ...",
 }): void {
-    if (reload_state.is_pending() || reload_state.is_in_progress()) {
+    if (reload_state.is_in_progress()) {
+        // If we're already attempting to reload, there's nothing to do.
+        return;
+    }
+
+    if (reload_state.is_pending() && !immediate) {
+        // If we're already pending and the caller is not requesting
+        // an immediate reload, there's nothing to do.
         return;
     }
 
@@ -218,6 +225,13 @@ export function initiate({
             server_reachable_check_failures = 0;
             if (immediate) {
                 do_reload_app(send_after_reload, save_compose, message_html);
+                // We don't expect do_reload_app to return, but if it
+                // does, the fallthrough logic seems fine.
+            }
+
+            if (reload_state.is_pending()) {
+                // If we're already pending, don't set the timers a second time.
+                return;
             }
 
             reload_state.set_state_to_pending();
```

laurynmm commented on web/src/reload.ts:
```diff
@@ -200,76 +203,137 @@ export function initiate({
+            const retry_delay_secs = Math.max(backoff_delay_secs, rate_limit_delay_secs);
+            setTimeout(() => {
+                initiate({immediate, save_compose, send_after_reload, message_html});
+            }, retry_delay_secs * 1000);
```
This is basically the same as what's used in `get_events` for the retry delay.

timabbott replied:
OK. This may be something we find reasonable to extract as a `util.get_retry_backoff(xhr, failures) type of` function that returns a simple object with any state properties to save as well as the desired delay.

timabbott replied:
If you end up doing another pass on this, let's do it; it's fairly likely we'll end up regretting forgetting to.

laurynmm replied:
Since I did another pass to fix the test failure, I added a commit to extract the delay backoff algorithms into the suggested helper function, but not the other state properties.

timabbott commented:
This seems to be broadly what we'd had in mind, so we can test-deploy this to see if it fixes the problem as intended.

laurynmm commented:
**Updates**:
- Added a prep commit that updates the selectors for the navigation puppeteer test.
  - This test was failing nondeterministically (could not be reproduced locally looping 100 times in the dev environment, but was failing on GitHub CI) after the changes to reload.initiate in this pull request.
  - Previously the test was waiting for the `message_feed_container` to be visible for a number of views, but that's true for a channel narrow, direct message view and combined feed view.
  - Updated the test to wait for the specific zulip-icon that's used in the message view header, which is what the direct message view test was doing.

timabbott commented:
Included this in today's chat.zulip.org deployment.

timabbott commented on web/src/reload.ts:
```diff
@@ -17,6 +18,8 @@ import * as util from "./util.ts";
 
 // Read https://zulip.readthedocs.io/en/latest/subsystems/hashchange-system.html
 
+let server_settings_failures = 0;
```
I renamed this and also switched to a cheaper API endpoint suggested by @alexmv.

``` diff
diff --git a/web/src/reload.ts b/web/src/reload.ts
index d1d02454b0..3c61389220 100644
--- a/web/src/reload.ts
+++ b/web/src/reload.ts
@@ -18,7 +18,7 @@ import * as util from "./util.ts";
 
 // Read https://zulip.readthedocs.io/en/latest/subsystems/hashchange-system.html
 
-let server_settings_failures = 0;
+let server_reachable_check_failures = 0;
 
 const token_metadata_schema = z.object({url: z.string(), timestamp: z.number()});
 
@@ -207,14 +207,15 @@ export function initiate({
         return;
     }
 
-    // In order to avoid races with the user's device connecting to the
-    // internet due to a network change while the device was suspended,
-    // we ping the server_settings API endpoint and initiate the reload
-    // once we get a success response from the server.
+    // In order to avoid races with the user's device connecting to
+    // the internet due to a network change while the device was
+    // suspended, we fetch a cheap unauthenticated API endpoint and
+    // only initiate the reload if we get a success response from the
+    // server.
     void channel.get({
-        url: "/api/v1/server_settings",
+        url: "/api/v1/compatibility",
         success() {
-            server_settings_failures = 0;
+            server_reachable_check_failures = 0;
             if (immediate) {
                 do_reload_app(send_after_reload, save_compose, message_html);
             }
@@ -312,8 +313,8 @@ export function initiate({
             // So we do the maximum of the retry-after header and an exponential
             // backoff with ratio sqrt(2) and half jitter. Starts at 1-2s and ends at
             // 45-90s after enough failures.
-            server_settings_failures += 1;
-            const backoff_scale = Math.min(2 ** ((server_settings_failures + 1) / 2), 90);
+            server_reachable_check_failures += 1;
+            const backoff_scale = Math.min(2 ** ((server_reachable_check_failures + 1) / 2), 90);
             const backoff_delay_secs = ((1 + Math.random()) / 2) * backoff_scale;
             let rate_limit_delay_secs = 0;
             const rate_limited_error_schema = z.object({
```

timabbott commented:
I pushed a couple tweaks and marked this to merge once CI passes again. Thanks @laurynmm! I think this will be a quite impactful improvement.

timabbott commented:
Hmm, that puppeteer test failed after a rerun, so I think there's something to debug there.

laurynmm commented:
**Updates**:
- Turned off auto-merge because these updates will need another review.
- Failing puppeteer test was due to the `channel.get` url being set to `/api/v1/compatibility` and returning a 404 response.
  - Changed it to `/compatibility`, but I wanted to make sure that's correct.
  - When I checked where [the mobile ReactNative app calls this endpoint](https://github.com/zulip/zulip-mobile/blob/7519372c2d78e87c8cd908f4f1bceaa99f8c3d81/src/api/checkCompatibility.js#L7), they call `https://zulip.com/compatibility`, but that didn't seem to be what we would want here.
- Added a commit to extract the exponential backoff algorithms used in `message_fetch.ts` and `server_events.js` into a helper function in `util.ts` as suggested above.

timabbott commented:
Yeah, this is great, merged, thanks @laurynmm! I had just made a typo in the URL.
